File: course_3/module_1/programming_assignment_1/solution_1.py
import sys


class Job:
    """
    Represents a job with a weight and length.
    """

    def __init__(self, weight, length):
        self.weight = weight
        self.length = length

    @property
    def weight_length_difference(self):
        """
        Calculates the difference (weight - length) for sorting.
        """
        return self.weight - self.length

    def __repr__(self):
        return f"Job(weight={self.weight}, length={self.length})"


class Scheduler:
    """
    Scheduler to minimize the weighted sum of completion times.
    """

    def __init__(self, jobs):
        """
        Initializes the scheduler with a list of jobs.
        :param jobs: List of Job objects.
        """
        self.jobs = jobs

    def get_sorted_jobs(self):
        """
        Returns a new list of jobs sorted based on the difference (weight - length).
        If two jobs have the same difference, prioritize by weight.
        :return: A sorted list of jobs.
        """
        return sorted(
            self.jobs,
            key=lambda job: (job.weight_length_difference, job.weight),
            reverse=True
        )

    def calculate_weighted_completion_time(self, sorted_jobs):
        """
        Calculates the weighted sum of completion times for the scheduled jobs.
        :param sorted_jobs: A list of jobs in the scheduled order.
        :return: The weighted sum of completion times as an integer.
        """
        total_time = 0
        completion_time = 0

        for job in sorted_jobs:
            completion_time += job.length
            total_time += job.weight * completion_time

        return total_time


def read_jobs_from_file(file_path):
    """
    Reads jobs from a file and creates Job objects.
    :param file_path: Path to the input file.
    :return: A list of Job objects.
    """
    jobs = []
    try:
        with open(file_path, "r") as file:
            lines = file.readlines()
            if not lines:
                raise ValueError("Input file is empty.")

            num_jobs = int(lines[0].strip())
            if num_jobs == 0:
                raise ValueError("No jobs specified in the input file.")

            for line in lines[1:]:
                try:
                    weight, length = map(int, line.split())
                    jobs.append(Job(weight, length))
                except ValueError:
                    print(f"Invalid data format in line: {line.strip()}")
        return jobs
    except FileNotFoundError:
        print(f"File not found: {file_path}")
        sys.exit(1)
    except ValueError as e:
        print(f"Error reading file: {e}")
        sys.exit(1)


if __name__ == "__main__":
    # Accept input file path as a command-line argument.
    file_path = sys.argv[1] if len(sys.argv) > 1 else "jobs.txt"

    # Read jobs from the input file.
    jobs = read_jobs_from_file(file_path)

    # Initialize the scheduler.
    scheduler = Scheduler(jobs)

    # Schedule jobs and calculate the total weighted completion time.
    sorted_jobs = scheduler.get_sorted_jobs()
    total_weighted_completion_time = scheduler.calculate_weighted_completion_time(sorted_jobs)

    # Output the result
    print(f"Total Weighted Completion Time: {total_weighted_completion_time}")

File: course_3/module_1/programming_assignment_1/solution_1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

// Job represents a job with weight and length.
type Job struct {
	Weight int
	Length int
}

// WeightLengthDifference calculates the difference (weight - length).
func (j Job) WeightLengthDifference() int {
	return j.Weight - j.Length
}

// Scheduler handles scheduling jobs to minimize weighted completion time.
type Scheduler struct {
	Jobs []Job
}

// NewScheduler creates a new Scheduler instance.
func NewScheduler(jobs []Job) *Scheduler {
	return &Scheduler{Jobs: jobs}
}

// ScheduleJobs sorts jobs based on the difference (weight - length).
// If two jobs have the same difference, it prioritizes by weight.
func (s *Scheduler) ScheduleJobs() {
	sort.Slice(s.Jobs, func(i, j int) bool {
		diff1 := s.Jobs[i].WeightLengthDifference()
		diff2 := s.Jobs[j].WeightLengthDifference()
		if diff1 == diff2 {
			return s.Jobs[i].Weight > s.Jobs[j].Weight
		}
		return diff1 > diff2
	})
}

// CalculateWeightedCompletionTime calculates the total weighted completion time.
func (s *Scheduler) CalculateWeightedCompletionTime() int {
	totalTime := 0
	completionTime := 0

	for _, job := range s.Jobs {
		completionTime += job.Length
		totalTime += job.Weight * completionTime
	}

	return totalTime
}

// ReadJobsFromFile reads jobs from a file and returns a list of Job objects.
func ReadJobsFromFile(filePath string) ([]Job, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	var jobs []Job
	scanner := bufio.NewScanner(file)
	var lineNum int

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		lineNum++

		// Skip the first line, which contains the number of jobs.
		if lineNum == 1 {
			continue
		}

		// Parse job details from the line.
		parts := strings.Fields(line)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid line format at line %d: %s", lineNum, line)
		}

		weight, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("invalid weight at line %d: %s", lineNum, parts[0])
		}

		length, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("invalid length at line %d: %s", lineNum, parts[1])
		}

		// Add the job to the list.
		jobs = append(jobs, Job{Weight: weight, Length: length})
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("error reading file: %w", err)
	}

	return jobs, nil
}

func main() {
	// Define the input file path.
	filePath := "course_3/module_1/programming_assignment_1/jobs.txt"

	// Read jobs from the file.
	jobs, err := ReadJobsFromFile(filePath)
	if err != nil {
		log.Fatalf("Error: %s\n", err)
	}

	// Ensure we have jobs to process.
	if len(jobs) == 0 {
		log.Fatalf("Error: No jobs found in the file.")
	}

	// Initialize the scheduler.
	scheduler := NewScheduler(jobs)

	// Schedule jobs and calculate the weighted completion time.
	scheduler.ScheduleJobs()
	totalWeightedCompletionTime := scheduler.CalculateWeightedCompletionTime()

	// Print the result.
	fmt.Printf("Total Weighted Completion Time: %d\n", totalWeightedCompletionTime)
}
